timesketch/lib/experimental/similarity.py

```diff
@@ -0,0 +1,204 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Calculate similarity scores based on the Jaccard distance between events."""
+from __future__ import unicode_literals
+
+import re
+from flask import current_app
+from datasketch.minhash import MinHash
+from datasketch.lsh import MinHashLSH
+from timesketch.lib.datastores.elastic import ElasticsearchDataStore
+
+
+class SimilarityScorerConfig(object):
+    """Configuration for a similarity scorer."""
+
+    # Parameters for Jaccard and Minhash calculations.
+    DEFAULT_THRESHOLD = 0.5
+    DEFAULT_PERMUTATIONS = 128
+
+    DEFAULT_CONFIG = {
+        'field': 'message',
+        'delimiters': [' ', '-', '/'],
+        'threshold': DEFAULT_THRESHOLD,
+        'num_perm': DEFAULT_PERMUTATIONS
+    }
+
+    # For any data_type that need custom config parameters.
+    # TODO: Move this to its own file.
+    # TODO: Add stopwords boolean config parameter.
+    # TODO: Add remove_words boolean config parameter.
+    CONFIG_REGISTRY = {
+        'windows:evtx:record': {
+            'query': 'data_type:"windows:evtx:record"',
+            'field': 'message',
+            'delimiters': [' ', '-', '/'],
+            'threshold': DEFAULT_THRESHOLD,
+            'num_perm': DEFAULT_PERMUTATIONS
+        }
+    }
+
+    def __init__(self, index, data_type):
+        """Initializes a similarity scorer config.
+
+        Args:
+            index: Elasticsearch index name.
+            data_type: Name of the data_type.
+        """
+        self._index = index
+        self._data_type = data_type
+        for k, v in self._get_config().items():
+            setattr(self, k, v)
```

Comment:
I don't love this config design - it's really hard to follow what configuration options there are, and where they're set.

Reply:
Yeah, I agree. The approach I take here is to have the user (developer for now, before we break out the config) to configure data_type configs in the dict, and then initialise the config objects attributes from that.

Will be refactored in the future, but keeping for now.